fem/bilininteg_hcurl.cpp

```diff
@@ -1322,12 +1633,325 @@ static void PACurlCurlApply3D(const int D1D,
+                     const double wcDy = Gc[dy][qy];
+                     const double wy = (dy < D1D-1) ? Bo[dy][qy] : 0.0;
+
+                     for (int qx = 0; qx < Q1D; ++qx)
+                     {
+                        const double O11 = sop[0 + (6*qx) + (6*Q1D*qy)];
+                        const double O12 = sop[1 + (6*qx) + (6*Q1D*qy)];
+                        const double O13 = sop[2 + (6*qx) + (6*Q1D*qy)];
+                        const double O22 = sop[3 + (6*qx) + (6*Q1D*qy)];
+                        const double O23 = sop[4 + (6*qx) + (6*Q1D*qy)];
+                        const double O33 = sop[5 + (6*qx) + (6*Q1D*qy)];
+
+                        const double c1 = (O11 * curl[qy][qx][0]) + (O12 * curl[qy][qx][1]) +
+                                          (O13 * curl[qy][qx][2]);
+                        const double c2 = (O12 * curl[qy][qx][0]) + (O22 * curl[qy][qx][1]) +
+                                          (O23 * curl[qy][qx][2]);
+                        const double c3 = (O13 * curl[qy][qx][0]) + (O23 * curl[qy][qx][1]) +
+                                          (O33 * curl[qy][qx][2]);
+
+                        const double wcx = Bc[dx][qx];
+                        const double wDx = Gc[dx][qx];
+
+                        if (dx < D1D-1)
+                        {
+                           // \hat{\nabla}\times\hat{u} is [0, (u_0)_{x_2}, -(u_0)_{x_1}]
+                           // (u_0)_{x_2} * (op * curl)_1 - (u_0)_{x_1} * (op * curl)_2
+                           const double wx = Bo[dx][qx];
+                           dxyz1 += (wx * c2 * wcy * wcDz) - (wx * c3 * wcDy * wcz);
+                        }
+
+                        // \hat{\nabla}\times\hat{u} is [-(u_1)_{x_2}, 0, (u_1)_{x_0}]
+                        // -(u_1)_{x_2} * (op * curl)_0 + (u_1)_{x_0} * (op * curl)_2
+                        dxyz2 += (-wy * c1 * wcx * wcDz) + (wy * c3 * wDx * wcz);
+
+                        // \hat{\nabla}\times\hat{u} is [(u_2)_{x_1}, -(u_2)_{x_0}, 0]
+                        // (u_2)_{x_1} * (op * curl)_0 - (u_2)_{x_0} * (op * curl)_1
+                        dxyz3 += (wcDy * wz * c1 * wcx) - (wcy * wz * c2 * wDx);
+                     } // qx
+                  } // qy
+               } // dx
+            } // dy
+         } // dz
+
+         MFEM_SYNC_THREAD;
+
+         MFEM_FOREACH_THREAD(dz,z,D1D)
+         {
+            MFEM_FOREACH_THREAD(dy,y,D1D)
+            {
+               MFEM_FOREACH_THREAD(dx,x,D1D)
+               {
+                  if (dx < D1D-1)
+                  {
+                     y(dx + ((dy + (dz * D1D)) * (D1D-1)), e) += dxyz1;
+                  }
+                  if (dy < D1D-1)
+                  {
+                     y(dx + ((dy + (dz * (D1D-1))) * D1D) + ((D1D-1)*D1D*D1D), e) += dxyz2;
+                  }
+                  if (dz < D1D-1)
+                  {
+                     y(dx + ((dy + (dz * D1D)) * D1D) + (2*(D1D-1)*D1D*D1D), e) += dxyz3;
```

Comment:
Same offset computation.

Reply:
In this case, for Nedelec elements, I do not see a way to use Reshape or DeviceTensor, because the dimensions are [D1D-1][D1D][D1D] for x, [D1D][D1D-1][D1D] for y, [D1D][D1D][D1D-1] for z.

Reply:
I agree we might need 3 of them

Reply:
@camierjs Using 3 `Reshape` instances would require using offsets for the y and z components, but offsets do not appear to be available yet. Maybe this can be added in the future.

Reply:
sure